### PR TITLE
[Keymap] @nstickney's layout for iris

### PR DESCRIPTION
**Pull request by nstickney**
## Description

Updated personal layout for ErgoDox; added personal layout for Iris.

## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [ ] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR
* 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by nstickney** [comment]
@drashna Wow! Didn't expect to get such great feedback on a personal keymap - thanks a million! I'll work through these to make sure I understand them and then resubmit, hopefully tonight.

**Comment by drashna** [comment]
Well, we try to make sure everything looks good.
And the rgblight eeprom stuff iis something I'm very familiar with. :)

**Comment by nstickney** [comment]
I find I'm having trouble understanding just this: how do the rgb lights get set in the eeprom the first time? As in the default color that should turn on when the keyboard boots?

**Comment by drashna** [comment]
Any time you run one of the rgblight commands that doesn't have the "noeeprom" part, it updates the EEPROM.  So, every time you start the keyboard, or a layer changes, it will write to the EEPROM. 

This may not seem like a lot, but it adds up very quickly, and may shorten the longevity of your controller. 

As for the "startup" value, I don't have a good answer for you there...
The best way is to use `matrix_scan_user` and a boolean to run a the initial setting on the first matrix scan.
Otherwise, the rgblight init function runs AFTER the matrix init, and will reset the rgblight settings, unless you write to EEPROM (eg, don't use a noeeprom function). 

basically, the startup is a mess until #3113 is merged in.

**Comment by nstickney** [comment]
@drashna weirdly, my backlighting now starts up in the FCTN color - I'm guessing this is something written to EEPROM. I tried rewriting the default color to EEPROM, but it didn't seem to have any effect.
Any thoughts?

**Comment by drashna** [comment]
Well, you can use the rgb keycodes to fix that.

Otherwise, you can use the first scan of `matrix_scan_user` to res the rgb color.  That's what I do personally. 
Eg:
```c
void matrix_scan_user(void) {
  static bool has_run;
  if (!has_run) {
    rgblight_enable_noeeprom();
    layer_state_set_user(layer_state);
    has_run = true;
  }
}
```

This should only run on the first cycle, and "never" again.   And it trigger's the `layer_state_set_user` function to set the RGB colors. 

If you're not using that, then you would want to replace the line with whatever code. 


Alternatively, I use this trick:
https://github.com/qmk/qmk_firmware/blob/master/users/drashna/rgb_stuff.c#L331-L365

**Comment by nstickney** [comment]
@drashna Thanks again for everything, I have it all working as desired now. Like I said before, super cool to get such great feedback on a personal layout. Submitted now for approval.

File: keyboards/iris/keymaps/nstickney/keymap.c [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+// Layers
+#define BASE 0 // Base layer
+#define SYMB 1 // Symbols
+#define SYSH 2 // Symbols, shifted
+#define NUMP 4 // Numpad
+#define FCTN 8 // Functions
+
+// Unicode input
+#undef UNICODE_SELECTED_MODES
+#define UNICODE_SELECTED_MODES UC_OSX, UC_LNX, UC_WINC
```
Comment (drashna) [inline keyboards/iris/keymaps/nstickney/keymap.c]: The undef and define should actually go in the `config.h` file, or they won't be picked up properly.

File: keyboards/iris/keymaps/nstickney/keymap.c [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+// Hack to properly initialize rgblight
+bool rgblight_initialized = false;
+void matrix_scan_user(void) {
+	if (!rgblight_initialized) {
+		rgblight_enable_noeeprom();
+		int i;
+		for ( i = 360; i > 18; i = i - 1) {
+			rgblight_sethsv(i, 255, 255);
```
Comment (drashna) [inline keyboards/iris/keymaps/nstickney/keymap.c]: Since you definitely don't want to write to eeprom for each of these....

```suggestion
			rgblight_sethsv_noeeprom(i, 255, 255);
```

File: layouts/community/ergodox/nstickney/keymap.c [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+ */
+
+// Layers
+#define BASE 0 // Base layer
+#define SYMB 1 // Symbols
+#define SYSH 2 // Symbols, shifted
+#define NUMP 4 // Numpad
+#define FCTN 8 // Function
+
+bool _layer_active(uint8_t layer) {
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]: Actually, you could use `IS_LAYER_ON` 

https://github.com/qmk/qmk_firmware/blob/c5221fa1cb1e903600205ba831c841f9d3aad33f/quantum/quantum.h#L220

File: layouts/community/ergodox/nstickney/keymap.c [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	//set_unicode_input_mode(UC_WIN); // Windows (with registry key, see wiki)
+	//set_unicode_input_mode(UC_WINC); // Windows (with WinCompose, see wiki)
+
+	ergodox_board_led_off();
+};
+
+// Runs constantly in the background, in a loop.
+void matrix_scan_user(void) {
+
+	if (_layer_active(SYMB) || _layer_active(SYSH)) {
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]:
```suggestion
	if (IS_LAYER_ON(SYMB) || IS_LAYER_ON(SYSH)) {
```

File: layouts/community/ergodox/nstickney/keymap.c [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+		ergodox_right_led_1_on();
+	} else {
+		ergodox_right_led_1_off();
+	}
+
+	if (_layer_active(NUMP)) {
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]:
```suggestion
	if (IS_LAYER_ON(NUMP)) {
```

File: layouts/community/ergodox/nstickney/keymap.c [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		ergodox_right_led_2_on();
+	} else {
+		ergodox_right_led_2_off();
+	}
+
+	if (_layer_active(FCTN)) {
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]:
```suggestion
	if (IS_LAYER_ON(FCTN)) {
```

File: keyboards/iris/keymaps/nstickney/keymap.c (hunk 2 above)
Comment (nstickney) [reply keyboards/iris/keymaps/nstickney/keymap.c]: :man_facepalming: I'm an idiot

File: layouts/community/ergodox/nstickney/keymap.c (hunk 3 above)
Comment (nstickney) [reply layouts/community/ergodox/nstickney/keymap.c]: I figured there was something like this available but it looks like I was searching the wrong things to find it.

File: layouts/community/ergodox/nstickney/keymap.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+#define SYMB 1 // Symbols
+#define SYSH 2 // Symbols, shifted
+#define NUMP 4 // Numpad
+#define FCTN 8 // Function
+
+// Tap Dancing
+void dance_lock (qk_tap_dance_state_t *state, void *user_data) {
+	switch (state->count){
+		case 1: // Press once for LGUI
+			register_code(KC_LGUI);
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]: You could actually use `tap_code` for all of these.  It does the same thing, but with an optional delay.

File: layouts/community/ergodox/nstickney/keymap.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		__________, KC_F12,
+		__________,
+		__________, __________, __________
+	),
+};
+
+// Runs just one time when the keyboard initializes.
+void matrix_init_user(void) {
+
+	set_unicode_input_mode(UC_LNX); // Linux
```
Comment (drashna) [inline layouts/community/ergodox/nstickney/keymap.c]: While this is fine here.... you could actually move this to `eeconfig_init_user` instead.  That way, it's not running every time. 

(but AVR code won't update it if it's the same value, so it will not do any harm)

File: layouts/community/ergodox/nstickney/keymap.c (hunk 7 above)
Comment (drashna) [reply layouts/community/ergodox/nstickney/keymap.c]: https://docs.qmk.fm/#/feature_macros?id=tap_codeltkcgt

Comment (nstickney) [reply layouts/community/ergodox/nstickney/keymap.c]: :+1:

File: layouts/community/ergodox/nstickney/keymap.c (hunk 8 above)
Comment (nstickney) [reply layouts/community/ergodox/nstickney/keymap.c]: Better to do things the best way. Thanks!

File: keyboards/iris/keymaps/nstickney/keymap.c [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+	_______,   KC_MPRV,   KC_MRWD,   KC_MPLY,   KC_MFFD,   KC_MNXT,   KC_F11,             KC_F12,    KC_MSTP,   KC_MUTE,   KC_VOLD,   KC_VOLU,   KC_BRID,   _______,  
+// └──────────┴──────────┴──────────┴────┬─────┴────┬─────┴────┬─────┴────┬─────┘        └────┬─────┴────┬─────┴────┬─────┴────┬─────┴──────────┴──────────┴──────────┘
+										  _______,   _______,   _______,                       _______,   _______,   _______
+//                                       └──────────┴──────────┴──────────┘                   └──────────┴──────────┴──────────┘
+	)
+};
+
+// Hack to properly initialize rgblight
+bool rgblight_initialized = false;
+void matrix_scan_user(void) {
```
Comment (drashna) [inline keyboards/iris/keymaps/nstickney/keymap.c]: Woooo, no longer need to do this. 
https://docs.qmk.fm/#/custom_quantum_functions?id=keyboard-post-initialization-code
(this was merged today).

You can use:
```c
void keyboard_post_init_user(void) {
		rgblight_enable_noeeprom();
		for (int i = 360; i > 0; i = i - 1) {
			rgblight_sethsv_noeeprom(i, 255, 255);
		}
		layer_state_set_user(layer_state);
}
```

**Comment by nstickney** [comment]
Just force-pushing to keep this updated with `master` for now until it gets merged. Thanks for all the help!

**Comment by drashna** [comment]
Awesome, thanks!

**Comment by drashna** [comment]
I really like that rgb init thing.  Mind if "borrow" it? 

Also, I find that it works better if you add a wait to it. Eg:
```c
void keyboard_post_init_user(void) {
	rgblight_enable_noeeprom();
	rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
	for (uint16_t i = 360; i > 0; i--) {
		rgblight_sethsv_noeeprom(i, 255, 255);
	}
	layer_state_set_user(layer_state);
}
```

**Comment by nstickney** [comment]
> I really like that rgb init thing. Mind if "borrow" it?
Sure! It ought to be a default initialization. :laughing:

> ```c
> 		wait_ms(5);
> ```
I was actually going to go digging for this but I haven't had a chance yet!

**Comment by drashna** [comment]
Just wanted to let you know that I improved this for my keymaps (which use the underglow for layer indication. 

```c
void keyboard_post_init_rgb(void) {
#ifdef RGBLIGHT_ENABLE
	rgblight_enable_noeeprom();
	layer_state_set_user(layer_state);
  uint16_t old_hue = rgblight_config.hue;
	rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
	for (uint16_t i = 360; i > 0; i--) {
		rgblight_sethsv_noeeprom( ( i + old_hue) % 360, 255, 255);
    wait_ms(8);
	}
	layer_state_set_user(layer_state);
#endif
}
```

This gets the default color for the active layer, saves it, and then starts and ends the cycle at that color.  So, it's a bit more customized. :)

**Comment by nstickney** [comment]
@drashna Sorry to bother you; I am wondering if a recent change affects the behavior of rgb_light. My lights no longer get all the way through their initial animation --- they stop on a ghastly yellow color --- and the color-per-layer code doesn't do anything. Am I doing something wrong? Is my board broken? Or is there something in QMK that's changed? I ask you because I can't find anything in the docs and you seemed to be about the only one who was interested in this part of the code.
